main.js

```diff
@@ -168,413 +30,609 @@ function getNodeRedEditorPath() {
         if (!fs.existsSync(nodeRedEditor)) {
             nodeRedEditor = path.normalize(`${__dirname}/../node_modules/@node-red/editor-client`);
             if (!fs.existsSync(nodeRedEditor)) {
-                adapter && adapter.log && adapter.log.error('Cannot find @node-red/editor-client packet!');
+                //adapter && adapter.log && adapter.log.error('Cannot find @node-red/editor-client packet!');
                 throw new Error('Cannot find @node-red/editor-client packet!');
             }
         }
     }
     return nodeRedEditor;
 }
 
-let redProcess;
-let stopping;
-let notificationsFlows;
-let notificationsCreds;
-let saveTimer;
 const nodePath = getNodeRedPath();
 const editorClientPath = getNodeRedEditorPath();
 
-function startNodeRed() {
-    adapter.config.maxMemory = parseInt(adapter.config.maxMemory, 10) || 128;
-    const args = [`--max-old-space-size=${adapter.config.maxMemory}`, path.join(nodePath, 'red.js'), '-v', '--settings', path.join(userDataDir, 'settings.js')];
+class NodeRed extends utils.Adapter {
+    constructor(options) {
+        super({
+            ...options,
+            name: 'node-red',
+            systemConfig: true,
+        });
 
-    if (adapter.config.safeMode) {
-        args.push('--safe');
+        this.systemSecret = null;
+        this.userDataDir = `${__dirname}/userdata/`;
+        this.redProcess = null;
+
+        this.stopping = false;
+        this.saveTimer = null;
+
+        this.notificationsFlows = null;
+        this.notificationsCreds = null;
+
+        this.attempts = {};
+        this.additional = [];
+
+        this.on('ready', this.onReady.bind(this));
+        //this.on('stateChange', this.onStateChange.bind(this));
+        this.on('message', this.onMessage.bind(this));
+        this.on('unload', this.onUnload.bind(this));
     }
 
-    adapter.log.info(`Starting node-red: ${args.join(' ')}`);
+    async onReady() {
+        this.installLibraries(() => {
+            if (this.config.projectsEnabled === undefined) this.config.projectsEnabled = false;
+            if (this.config.allowCreationOfForeignObjects === undefined) this.config.allowCreationOfForeignObjects = false;
 
-    redProcess = spawn('node', args);
+            // create userData directory
+            if (!fs.existsSync(this.userDataDir)) {
+                fs.mkdirSync(this.userDataDir);
+            }
 
-    redProcess.on('error', err =>
-        adapter.log.error(`catched exception from node-red:${JSON.stringify(err)}`));
+            this.generateHtml()
+                .then(() => {
```

Comment:
nit: why not using await here :-)

Reply:
That was the original code - not everything was rewritten 😄